REQUIREMENTS.md

```diff
@@ -0,0 +1,160 @@
+*every* IPLD implementation needs to support these new formats.
+
+## No Non-Local Reasoning
+
+An IPLD block should never be interpreted in the context of *anything* not
+contained in the block (and CID).
+
+For example, assuming we add support for relative links, the following
+definition of `foo` would not be a valid IPLD block:
+
+```
+var foo = {
+  // points outside of the current block, into the parent's "baz" field.
+  "baz": {"/": "../../baz"}
+}
+var bar = {
+  "foo": CidOf(foo),
+  // `/foo/baz` points here.
+  "baz": "something"
+}
+
+// resolution throug block `foo` depends on block `bar`.
+Resolve("/ipld/${CidOf(bar)}/foo/baz/")
+```
+
+For the same reason, IPLD links can't rely on an authority (e.g., a blockchain).
+
+Note: Links like this can still be encoded at the application level but they
+won't be handled by the IPLD resolver (and won't get the special "link" type).
+
+**Motivation:** IPLD needs to be easy to reason about.
+
+**Caveat:**
+
+We *may* want to relax this if we want to move schemas into separate,
+deduplicated blocks (referenced by CID). If we do that, we'd need to fetch a
+block's schema before being able to interpret the it.
+
+However, we need to *thoroughly* discuss any changes to this requirement.
+
+1. The space savings may not be worth it given the size of CIDs (>40 bytes),
+   compression, smart transports, and smart datastores.
+2. This change would introduce some weird interface complexities and potential
+   network dependencies.
+
+## No Cycles
+
+IPLD links must not be cyclic, even if we add support for relative links.
+
+**Motivation:**
+
+1. Security: Forbidding link cycles ensures that any graph traversal terminates.
+   This makes it easier to correctly and securely implement some graph
+   algorithms, even on potentially untrusted data in a distributed system.
+2. Consistency: Without complex hacks, it's impossible to create link cycles
+   *between* hash-linked blocks. Allowing link cycles one one level (e.g., in
+   relative links within a single block) but not on a larger scale (between
+   blocks) is inconsistent (even though there's nothing we can do about it).
+
+## Stable Pathing
+
+An IPLD path always means the same thing, everywhere, every time. Importantly,
+users shouldn't be able to configure their IPLD library to change how path
+resolution works. They should be able to use alternative path resolution
+algorithms for *non*-IPLD paths (e.g., IPFS paths) but those should be built
+on top of IPLD.
+
+**Motivation:** Deterministic computations on top of a IPLD need to produce the
+same result every time.
+
+## Link Transparent Pathing
+
+Path resolution must transparently traverse links.
+
+**Motivation:** Deduplication and composability. If links are *not* transparent,
+programmers will have to either make many small objects (lots of hashing) or
+inline data into large objects (lots of duplication and copying).
+
+## Primitives
```

Comment:
We could refer to the [Data Model spec](https://github.com/ipld/specs/blob/8d7dbe757a1b42c5bb3704d29cf2fffdf8639d4d/data-model-layer/data-model.md).